src/wrapInTransaction.ts

```diff
@@ -0,0 +1,133 @@
+
+    const runOriginal = async () => fn.apply(this, [...newArgs])
+    const runWithNewHook = async () => runInNewHookContext(context, runOriginal)
+
+    const runWithNewTransaction = async () => {
+      const transactionCallback = async (entityManager: EntityManager) => {
+        log(
+          `runWithNewTransaction - set entityManager in context: isCurrentTransactionActive: ${entityManager?.queryRunner?.isTransactionActive}`
+        )
+        setEntityManagerForConnection(connectionName, context, entityManager)
+        try {
+          const result = await fn.apply(this, [...newArgs])
+          log(`runWithNewTransaction - Success`)
+          return result
+        } catch (e) {
+          log(`runWithNewTransaction - ERROR|${e}`)
+          throw e
+        } finally {
+          log(`runWithNewTransaction - reset entityManager in context`)
+          setEntityManagerForConnection(connectionName, context, null)
+        }
+      }
+
+      if (isolationLevel) {
+        return await runInNewHookContext(context, () =>
+          getManager(connectionName).transaction(isolationLevel, transactionCallback)
+        )
+      } else {
+        return await runInNewHookContext(context, () =>
+          getManager(connectionName).transaction(transactionCallback)
+        )
+      }
+    }
+
+    return context.runAndReturn(async () => {
+      const currentTransaction = getEntityManagerForConnection(connectionName, context)
+
+      switch (propagation) {
+        case Propagation.MANDATORY:
+          if (!currentTransaction) {
+            throw new TransactionalError(
+              "No existing transaction found for transaction marked with propagation 'MANDATORY'"
+            )
+          }
+          return runOriginal()
+        case Propagation.NESTED:
+          return runWithNewTransaction()
+        case Propagation.NEVER:
+          if (currentTransaction) {
+            throw new TransactionalError(
+              "Found an existing transaction, transaction marked with propagation 'NEVER'"
+            )
+          }
+          return runWithNewHook()
+        case Propagation.NOT_SUPPORTED:
+          if (currentTransaction) {
+            setEntityManagerForConnection(connectionName, context, null)
+            const result = await runWithNewHook()
+            setEntityManagerForConnection(connectionName, context, currentTransaction)
+            return result
+          }
+          return runOriginal()
+        case Propagation.REQUIRED:
+          if (currentTransaction) {
+            return runOriginal()
+          }
+          return runWithNewTransaction()
+        case Propagation.REQUIRES_NEW:
+          return runWithNewTransaction()
+        case Propagation.SUPPORTS:
+          if (currentTransaction) {
+            return runOriginal()
+          } else {
+            return runWithNewHook()
+          }
+      }
+    }) as unknown as ReturnType<Func>;
+  }
+
+  return wrapped as Func;
+}
```

Comment:
Please [add a newline](https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file) here...